src/merkle_tree.rs

```diff
@@ -0,0 +1,488 @@
+                if index + 1 < level.len() as u32 {
+                    proof.push((level[(index + 1) as usize].clone(), Direction::Right));
+                } else {
+                    proof.push((level[index as usize].clone(), Direction::Right));
+                }
+            } else {
+                proof.push((level[(index - 1) as usize].clone(), Direction::Left));
+            }
+
+            index /= 2;
+        }
+
+        Ok(ProofOfInclusion::new_from(leaf, proof))
+    }
+
+    /// Returns a proof of inclusion for a given hash in the Merkle Tree. The proof generated conains the hashes of the siblings of the nodes in the path from the leaf to the root, and their directions. In O(n) time, with n = number of leaf hashes.
+    ///
+    /// # Parameters
+    /// - `leaf`: The hash to generate the proof for
+    ///
+    /// # Returns
+    /// A Result that, if the hash given is included in the tree, contains a `ProofOfInclusion` containing the proof of inclusion for the given hash. If the hash is not included in the tree, an error message is returned.
+    pub fn proof_of_inclusion(&self, leaf: MerkleHash) -> Result<ProofOfInclusion, &str> {
+        let hash_index = match self.levels[0].iter().position(|h| h == &leaf) {
+            Some(index) => index,
+            None => return Err("Hash not found in tree"),
+        };
+
+        self.proof_of_inclusion_with_index(leaf, hash_index as u32)
+    }
+
+    /// Adds a hash to the Merkle Tree, updating the tree structure.
+    ///
+    /// # Parameters
+    /// - `hash`: The hash to add to the tree
+    pub fn add_hash(&mut self, hash: MerkleHash) -> Result<(), &str> {
+        let len = self.levels[0].len();
+
+        if self.verify(hash.clone()) {
+            return Err("Hash is already in the tree");
+        }
+
+        if self.levels[0][len - 1] == self.levels[0][len - 2] {
+            self.levels[0][len - 1] = hash;
+        } else {
+            self.levels[0].push(hash);
+        }
+
+        let mut new_tree = MerkleTree { levels: vec![] };
+
+        MerkleTree::build_tree(&mut new_tree, self.levels[0].clone());
+
+        self.levels = new_tree.levels;
+        Ok(())
+    }
+
+    /// Adds an element that will be hashed before adding it to the Merkle Tree, .
+    pub fn add_data(&mut self, data: impl AsRef<[u8]>) -> Result<(), &str> {
+        let hash = MerkleTree::get_hash_of(&data);
+        self.add_hash(hash)
+    }
+
+    /// Converts a byte slice to a hexadecimal string.
+    pub fn bytes_to_hex(bytes: &[u8]) -> String {
+        let hex_chars: Vec<String> = bytes.iter().map(|byte| format!("{:02x}", byte)).collect();
+        hex_chars.join("")
+    }
+
+    /// Prints the Merkle Tree structure.
+    pub fn print(&self) {
+        for i in 0..self.levels.len() {
+            println!("LEVEL {}:", i);
+            for hash in self.levels[i].iter() {
+                println!("- {:?}", hash);
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Maybe it would be a good practice to place tests in another file instead of storing them in merkle_tree.rs, but the choice is ultimately yours

Reply:
I always thought that the convention in Rust was that unit tests go in the same file of the code they are testing, look [here](https://doc.rust-lang.org/book/ch11-03-test-organization.html)

Reply:
Good to know. Thanks for sharing, haven't read that before.